ui/src/shared/components/HistogramChartBars.tsx

```diff
@@ -1,27 +1,175 @@
-import React, {PureComponent} from 'react'
+import React, {PureComponent, MouseEvent} from 'react'
 import _ from 'lodash'
 import {ScaleLinear, ScaleTime} from 'd3-scale'
+import {color} from 'd3-color'
 
-import {HistogramData, HistogramDatum} from 'src/types/histogram'
+import {getDeep} from 'src/utils/wrappers'
+
+import {
+  HistogramData,
+  HistogramDatum,
+  HoverData,
+  TooltipAnchor,
+  ColorScale,
+} from 'src/types/histogram'
 
 const BAR_BORDER_RADIUS = 4
 const BAR_PADDING_SIDES = 4
+const HOVER_BRIGTHEN_FACTOR = 0.4
+const TOOLTIP_HORIZONTAL_MARGIN = 5
+const TOOLTIP_REFLECT_DIST = 100
+
+const getBarWidth = ({data, xScale, width}): number => {
+  const dataInView = data.filter(
+    d => xScale(d.time) >= 0 && xScale(d.time) <= width
+  )
+  const barCount = Object.values(_.groupBy(dataInView, 'time')).length
+
+  return Math.round(width / barCount - BAR_PADDING_SIDES)
+}
+
+type SortFn = (a: HistogramDatum, b: HistogramDatum) => number
+
+const getSortFn = (data: HistogramData): SortFn => {
+  const counts = {}
+
+  for (const d of data) {
+    if (counts[d.group]) {
+      counts[d.group] += d.value
+    } else {
+      counts[d.group] = d.value
+    }
+  }
+
+  return (a, b) => counts[b.group] - counts[a.group]
```

Comment:
Returning an object here that has this  as a method and `counts` as a property would save you from computing the sum again on line 93.